### PR TITLE
[veomni, cfg] feat: add missing config fields to veomni.yaml

### PR DESCRIPTION
**Pull request by mikequan0425**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

Supplement the configurations available in engine.py but missing in engine/veomni.yaml, and add new optional operators in the veomni OpsImplementationConfig.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+veomni+config
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

After completing the configuration in the script, the logs show that it has taken effect successfully.

image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.
Add new configuration items for the veomni training engine
```python
actor_rollout_ref.actor.veomni.entropy_from_logits_with_chunking=True \
actor_rollout_ref.actor.veomni.offload_policy=False
```

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/trainer/config/engine/veomni.yaml [hunk 1]
```diff
@@ -28,15 +40,31 @@ enable_full_shard: true
...
 # Only for FSDP1: FSDP1 configuration, prefetch the next forward-pass all-gather
 # before the current forward computation.
 forward_prefetch: true
 
+# Reshard after forward pass to reduce memory footprint
+# For FSDP1, `false` enables `ShardingStrategy.SHARD_GRAD_OP`
+reshard_after_forward: true
+
+# Whether to use original parameters in fsdp. Only available in fsdp1
+use_orig_params: false
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply verl/trainer/config/engine/veomni.yaml]: Agree

File: verl/trainer/config/engine/veomni.yaml [hunk 2]
```diff
@@ -16,6 +25,9 @@ expert_parallel_size: 1
 
 mixed_precision: true
 
+# Mixed precision training param dtype
+dtype: bfloat16
```
Comment (FoolPlayer) [inline verl/trainer/config/engine/veomni.yaml]: Agree with gemini review, VeOmni use fp32 optimizer when enable mixed_precision. If adding model dtype it's better add some assert or type casting when dtype=bf16 and mixed_precision=true

File: verl/trainer/config/engine/veomni.yaml [hunk 3]
```diff
@@ -7,6 +7,15 @@ param_offload: False
 # Whether to offload optimizer state to CPU
 optimizer_offload: False
 
+# Whether to offload gradients to CPU
+grad_offload: false
```
Comment (FoolPlayer) [inline verl/trainer/config/engine/veomni.yaml]: grad_offload is not support by veomni/fsdp2 now, veomni just use fsdp2's offload api `CPUOffloadPolicy()`. It not support setting seperately

File: verl/trainer/config/engine/veomni.yaml [hunk 4]
```diff
@@ -28,15 +40,31 @@ enable_full_shard: true
...
+# Path to load checkpoint from, if any
+load_checkpoint_path: null
+
 # Only for FSDP1: FSDP1 configuration, prefetch the next forward-pass all-gather
 # before the current forward computation.
 forward_prefetch: true
 
+# Reshard after forward pass to reduce memory footprint
+# For FSDP1, `false` enables `ShardingStrategy.SHARD_GRAD_OP`
+reshard_after_forward: true
```
Comment (FoolPlayer) [inline verl/trainer/config/engine/veomni.yaml]: Same for this arg, now veomni support FSDP2 only.

File: verl/trainer/config/engine/veomni.yaml [hunk 5]
```diff
@@ -7,6 +7,15 @@ param_offload: False
...
 optimizer_offload: False
 
+# Whether to offload gradients to CPU
+grad_offload: false
+
+# Only for FSDP2: offload param/grad/optimizer during train
+offload_policy: false
+
+# policy for wrapping the model
+wrap_policy: {}
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply verl/trainer/config/engine/veomni.yaml]: Agree

**Review by FoolPlayer — CHANGES_REQUESTED** [review]
Thanks for your PR, the only problem is some args only for FSDP1, but VeOmni has already remove FSDP1 support after v0.1.10. And VeOmni 0.1.11 will be used with verl 0.8.0

**Comment by mikequan0425** [comment]
@FoolPlayer Thank you for your review and your excellent work on Veomni. I have revised everything based on your suggestions.

**Review by FoolPlayer — DISMISSED** [review]

**Comment by mikequan0425** [comment]
@FoolPlayer @wuxibin89 pre-commit failed in CI, I've pushed the updates again. Could you help trigger CI again? Thanks!

File: verl/workers/config/engine.py [hunk 6]
```diff
@@ -272,11 +272,9 @@ class VeOmniEngineConfig(EngineConfig):
     The inheritance from BaseConfig provides omegaconf.DictConfig-like interface for a dataclass config.
 
     Args:
-        wrap_policy (Dict[str, Any]): Configuration for FSDP wrap policy.
         param_offload (bool): Whether to offload parameters to CPU, default False
         optimizer_offload (bool): Whether to offload optimizer states to CPU, default False
         offload_policy (bool): Whether to offload policy model parameters, default False
```
Comment (tardis-key) [inline verl/workers/config/engine.py]: remove offload_policy

Comment (mikequan0425) [reply verl/workers/config/engine.py]: done

**Review by Luosuu — APPROVED** [review]

**Comment by tardis-key** [comment]
There are lots of ci error about 'No matching distribution found for TransferQueue==0.1.8'. It seems that the PyPI mirror used by CI has a certain amount of time lag.
